File: src/ServerScriptService/Modules/GameLibrary/Survivor.lua
local TweenService = game:GetService("TweenService")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local Players = game:GetService("Players")
local ServerScriptService = game:GetService("ServerScriptService")
local Modules = ServerScriptService:WaitForChild("Modules")
local RModules = ReplicatedStorage:WaitForChild("Modules")
local RoleCommunication = require(RModules:WaitForChild("RoleCommunication"))
local DataCreator = require(Modules:WaitForChild("DataCreator"))
local ServerUtil = require(Modules:WaitForChild("ServerUtil"))
local Util = require(RModules:WaitForChild("Util"))
local SurvivorAnimations = ReplicatedStorage:WaitForChild("SurvivorAnimations")
local Sounds = ReplicatedStorage:WaitForChild("Sounds")

local survivors = DataCreator.newOrGet("List", "Survivors")

local survivor = {}

-- Makes the Survivor blink
function survivor:blink()
	local data : DataCreator.Data = self:getData()
	local track = self.blinkTrack :: AnimationTrack
	if(not data:getValue("dead")) then
		if(not data:getValue("blinking")) then
			self:queueStraining(false)
			data:setValue("strainTime", data:getValue("blinkResetTimer") / 2)
			self:setBlinkMeter(0) -- for the sake of eye animation
			self:resetValues()
			data:setValue("blinking", os.clock())
		else
			local delta = os.clock() - data:getValue("blinking")
			if(delta >= data:getValue("blinkResetTimer")) then
				data:setValue("blinking", nil)
				self:setBlinkMeter(data:getValue("blinkMeter")) -- for the sake of eye animation
				self:queueStraining(true)
			end
		end
	end
end

-- Start straining a Survivor's eyes
-- @param start : Whether to start or stop straining
function survivor:queueStraining(start : boolean)
	local data : DataCreator.Data = self:getData()

	if(data:getValue("dead")) then return end

	if(start) then
		data:setValue("straining", os.clock())
	else
		data:setValue("straining", nil)
	end
end

-- Lets the survivor strain their eyes for slightly longer, increasing the meter but also making their eyes strain faster
function survivor:strain()
	local data : DataCreator.Data = self:getData()

	if(data:getValue("dead")) then return end

	if data:getValue("blinkMeter") < data:getValue("maxBlinkMeter") and data:getValue("strainIncrease") > 0 and not data:getValue("blinking") then
		self:setBlinkMeter(data:getValue("blinkMeter") + data:getValue("strainIncrease"))
		data:setValue("strainIncrease", math.max(data:getValue("strainIncrease") - 0.6, data:getValue("minStrainIncrease")))
		data:setValue("strainTime", math.max(data:getValue("strainTime") - 0.04, data:getValue("minStrainTime")))
	end
end

-- Sets the blink meter to a new value. This also changes the face animation
-- @param newValue : The value to change to
function survivor:setBlinkMeter(newValue : number)
	local data : DataCreator.Data = self:getData()

	if(data:getValue("dead")) then return end

	data:setValue("blinkMeter", math.min(newValue, data:getValue("maxBlinkMeter")))
	local track = self.blinkTrack :: AnimationTrack
	local trackTime = math.min(1 - (newValue / data:getValue("maxBlinkMeter")), 0.99)
	--track.TimePosition = trackTime
	local useTime = (newValue == 100 and data:getValue("blinkResetTimer") / 2) or data:getValue("strainTime")
	local tInfo = TweenInfo.new(useTime, Enum.EasingStyle.Sine, Enum.EasingDirection.In, 0, false, 0)
	local tweenToPosition = TweenService:Create(track, tInfo, {TimePosition = trackTime})
	if(not track.IsPlaying) then
		track:Play()
		track:AdjustSpeed(0)
	end
	tweenToPosition:Play()
end

function survivor:isAlive()
	local data : DataCreator.Data = self:getData()
	return not data:getValue("dead")
end

-- Kills the survivor
function survivor:kill()
	local data : DataCreator.Data = self:getData()
	if(not data:getValue("dead")) then
		data:setValue("dead", true)
		ServerUtil.removeWatchingAngelsForCharacter(data:getObject())

		local character : Model = data:getObject()
		local clone = character:Clone()
		character:Destroy()
		local humanoid : Humanoid = clone:WaitForChild("Humanoid")
		-- avoid HRP colliding with UpperTorso on death
		--clone.HumanoidRootPart.CanCollide = false

		clone.Parent = workspace

		for _, instance : Instance in clone:GetDescendants() do
		--	if(instance:IsA("BasePart")) then
		--		--part:SetNetworkOwner(nil)
		--		instance.Anchored = false

		--		if(instance ~= clone.HumanoidRootPart and not instance.Parent:IsA("Accessory") and instance.Name ~= "Handle") then
		--			instance.CanCollide = true
		--		end
			if instance:IsA("BaseScript") then
				instance:Destroy()
		--	elseif instance:IsA("AnimationConstraint") then
		--		if(instance.Parent == clone.Head) then continue end

		--		instance.Enabled = false
		--	elseif instance:IsA("BallSocketConstraint") then
		--		--instance.MaxFrictionTorque = 1
			end
		end
		
		--humanoid:SetStateEnabled(Enum.HumanoidStateType.GettingUp, false)
		--humanoid:SetStateEnabled(Enum.HumanoidStateType.Jumping, false)
		--humanoid:ChangeState(Enum.HumanoidStateType.Dead)
		humanoid.Health = 0
		humanoid.AutoJumpEnabled = false

		data:setValue("ragdoll", clone)

		for _, track : AnimationTrack in humanoid.Animator:GetPlayingAnimationTracks() do
			track:Stop()
		end

		local track : AnimationTrack = humanoid.Animator:LoadAnimation(SurvivorAnimations.Dead:Clone())
		track:Play()
		track:AdjustSpeed(0)
		track.TimePosition = 0.99

		local snaps : {Sound} = Sounds.Snaps:GetChildren()
		local randomSnap = snaps[math.random(1, #snaps)]:Clone()
		randomSnap.Parent = clone.Head
		randomSnap:Play()

		local player : Player = data:getValue("player")
		if(player) then
			task.spawn(pcall, function()
				task.wait(Players.RespawnTime)
				if(not player.Character:IsDescendantOf(game)) then
					player:LoadCharacter()
				end
			end)
		end
	end
end

-- Removes the survivor
function survivor:remove()
	local data : DataCreator.Data = self:getData()
	survivors:setValue(data:getObject(), nil)
	local ragdoll : Model = data:getValue("ragdoll")
	if(ragdoll) then
		ragdoll:Destroy()
	end
	data:remove()
	table.freeze(self)
	self = nil
end

-- Reset values to their default, used when the Survivor blinks
function survivor:resetValues()
	local data : DataCreator.Data = self:getData()
	data:setValue("maxBlinkMeter", 100)
	data:setValue("blinkMeter", data:getValue("maxBlinkMeter"))
	data:setValue("blinkResetTimer", .3)

	data:setValue("drainAmount", 10)

	data:setValue("straining", nil)
	data:setValue("strainTime", 1)
	data:setValue("minStrainTime", 0.2)

	data:setValue("strainIncrease", 15)
	data:setValue("minStrainIncrease", 0)
end

-- Returns the data for this survivor
function survivor:getData() : DataCreator.Data
	return self.data
end

export type Survivor = typeof(survivor)

return survivor

File: src/server/Roles/Survivor.ts
import {
	PhysicsService,
	Players,
	ReplicatedStorage,
	ServerScriptService,
	TweenService,
	Workspace,
} from "@rbxts/services";
import { PlayerKeyPredicate, ServerDataObject } from "../ServerDataObject";
import { Actor } from "./Actor";
import { Array } from "@rbxts/luau-polyfill";
import { Util } from "shared/Util";
import EasyRagdoll from "shared/EasyRagdoll";

export const SurvivorList = ServerDataObject.getOrConstruct<string>("List", ["Survivor"]);

SurvivorList.setCriteriaForDataObject(() => true);
SurvivorList.setFutureCriteriaForKeys(() => {
	return { canSeeKey: true, canSeeValue: false, canEditValue: false };
});

export class Survivor extends Actor {
	public static getOrCreate(character: Model, player?: Player): Survivor {
		let survivor = SurvivorList.getValue<Survivor>(character);
		if (survivor !== undefined) return survivor;

		survivor = new Survivor(character, player);
		SurvivorList.setValue(character, survivor);

		return survivor;
	}

	private constructor(character: Instance, player?: Player) {
		super(ServerDataObject.getOrConstruct<Instance>(character, ["Survivor"]));

		const humanoid = character.WaitForChild("Humanoid") as Humanoid;
		let animator = humanoid.FindFirstChild("Animator") as Animator;
		if (animator === undefined) {
			animator = new Instance("Animator");
			animator.Parent = humanoid;
		}

		humanoid.HealthDisplayType = Enum.HumanoidHealthDisplayType.AlwaysOff;
		humanoid.DisplayDistanceType = Enum.HumanoidDisplayDistanceType.None;
		humanoid.UseJumpPower = true;
		humanoid.JumpPower = 0;
		humanoid.BreakJointsOnDeath = false;

		this.data.setValue("blinking", undefined);
		this.data.setValue("blinkTrack", animator.LoadAnimation(ReplicatedStorage.SurvivorAnimations.Blink.Clone()));

		if (player !== undefined) {
			this.data.setValue("player", player);

			const predicate: PlayerKeyPredicate = (_player: Player) => {
				return { canSeeKey: _player.UserId === player.UserId, canSeeValue: true, canEditValue: false };
			};

			this.data.setPlayerCriteriaForKeys(["blinking", "blinkMeter", "dead", "maxBlinkMeter"], predicate);
			this.data.setCriteriaForDataObject((_player) => _player.UserId === player.UserId);
		}

		character.Archivable = true;

		this.defaultValues();
		this.queueStraining(true);
	}

	public defaultValues() {
		this.data.setValue("maxBlinkMeter", 100);
		this.data.setValue("blinkMeter", this.data.getValue("maxBlinkMeter"));
		this.data.setValue("blinkResetTimer", 0.3);

		this.data.setValue("drainAmount", 10);

		this.data.setValue("straining", undefined);
		this.data.setValue("strainTime", 1);
		this.data.setValue("minStrainTime", 0.2);

		this.data.setValue("strainIncrease", 15);
		this.data.setValue("minStrainIncrease", 0);
	}

	public blink(): boolean {
		if (!this.isAlive()) return false;

		if (this.data.getValue<number>("blinking") !== undefined) return false;

		this.defaultValues();
		this.setBlinkMeter(0, true);
		this.data.setValue("blinking", os.clock());

		return true;
	}

	public queueStraining(start: boolean) {
		if (!this.isAlive()) return;

		this.data.setValue("straining", start ? os.clock() : undefined);
	}

	public setBlinkMeter(newValue: number, fastAnimation?: boolean) {
		if (!this.isAlive()) return;

		this.data.setValue("blinkMeter", math.min(newValue, this.data.getValue("maxBlinkMeter")));
		const track = this.data.getValue<AnimationTrack>("blinkTrack");
		const trackTime = math.min(1 - newValue / this.data.getValue<number>("maxBlinkMeter"), 0.99);
		const useTime = fastAnimation
			? this.data.getValue<number>("blinkResetTimer") / 2
			: this.data.getValue<number>("strainTime");
		const tInfo = new TweenInfo(useTime, Enum.EasingStyle.Sine, Enum.EasingDirection.In, 0, false, 0);
		const tweenToPosition = TweenService.Create(track, tInfo, { TimePosition: trackTime });
		if (!track.IsPlaying) {
			track.Play();
			track.AdjustSpeed(0);
		}
		tweenToPosition.Play();
	}

	public strain(): boolean {
		if (!this.isAlive()) return false;

		const blinkMeter = this.data.getValue<number>("blinkMeter");
		const strainIncrease = this.data.getValue<number>("strainIncrease");
		if (
			blinkMeter < this.data.getValue<number>("maxBlinkMeter") &&
			strainIncrease > 0 &&
			!this.data.getValue<boolean>("blinking")
		) {
			this.setBlinkMeter(blinkMeter + strainIncrease);
			this.data.setValue(
				"strainIncrease",
				math.max(strainIncrease - 0.6, this.data.getValue<number>("minStrainIncrease")),
			);
			this.data.setValue(
				"strainTime",
				math.max(this.data.getValue<number>("strainTime") - 0.04, this.data.getValue<number>("minStrainTime")),
			);
			return true;
		}
		return false;
	}

	public isAlive() {
		return !this.data.getValue<boolean>("dead");
	}

	public die() {
		super.die();
		const character = this.getData().getHolder();
		if (character !== undefined) {
			const clone = character.Clone() as Model;
			clone
				.GetDescendants()
				.filter((value) => value.IsA("BaseScript"))
				.forEach((value) => value.Destroy());
			clone.Parent = Workspace;
			clone.Name = "Ragdoll";

			character.Destroy();

			EasyRagdoll.SetRagdoll(clone, true, false);
			// clone
			// 	.GetChildren()
			// 	.filter(
			// 		(value) =>
			// 			value.Name !== "HumanoidRootPart" && value.Name !== "CollisionPart" && value.IsA("BasePart"),
			// 	)
			// 	.forEach((value) => {
			// 		if (value.IsA("BasePart")) {
			// 			value.CanCollide = true;
			// 			value.GetPropertyChangedSignal("CanCollide").Connect(() => (value.CanCollide = true));
			// 		}
			// 	});
			// (clone.WaitForChild("CollisionPart") as BasePart).CanCollide = false;
			// (clone.WaitForChild("HumanoidRootPart") as BasePart).CanCollide = false;
			clone
				.GetDescendants()
				.filter((value) => value.IsA("BasePart"))
				.forEach((value) => (value.CollisionGroup = "Ragdolls"));

			const player = this.data.getValue<Player>("player");
			if (player !== undefined) {
				task.spawn(() => {
					task.wait(Players.RespawnTime);
					player.LoadCharacter();
				});
			}
		}
		this.destroy();
	}

	public destroy() {
		SurvivorList.removeKey(this.getData().getHolder());
		super.destroy();
	}

	public tick() {
		super.tick();

		if (!this.isAlive()) return;

		const humanoid = this.data.getHolder().FindFirstChildWhichIsA("Humanoid");
		if (humanoid !== undefined) {
			humanoid.WalkSpeed = Util.getWalkDirection(humanoid).Y >= 0 ? 16 : 10;
		}

		const strainValue = this.data.getValue<number>("straining");
		const blinking = this.data.getValue<number>("blinking");
		if (strainValue !== undefined) {
			const deltaTime = os.clock() - strainValue;
			if (deltaTime > this.data.getValue<number>("strainTime")) {
				const newBlinkMeter = math.max(
					this.data.getValue<number>("blinkMeter") - this.data.getValue<number>("drainAmount"),
					0,
				);
				this.setBlinkMeter(newBlinkMeter);

				if (this.data.getValue<number>("blinkMeter") <= 0) {
					this.data.setValue("straining", undefined);
					this.blink();
				} else {
					this.data.setValue("straining", os.clock());
				}
			}
		} else if (blinking !== undefined) {
			// currently blinking
			const deltaTime = os.clock() - blinking;
			if (deltaTime >= this.data.getValue<number>("blinkResetTimer")) {
				this.data.setValue("blinking", undefined);
				this.setBlinkMeter(this.data.getValue("maxBlinkMeter"), true);
				this.queueStraining(true);
			}
		}
	}
}
